runner/main.go

```diff
@@ -0,0 +1,225 @@
+func Run(m *testing.M, exitOnError bool, failRetriesCount int) int {
+	runner := &testRunner{
+		m:                m,
+		exitOnError:      exitOnError,
+		failRetriesCount: failRetriesCount,
+	}
+	runner.init()
+	return runner.Run()
+}
+
+// Initialize test runner, replace the internal test with an indirection
+func (r *testRunner) init() {
+	if tPointer, err := getFieldPointerOfM(r.m, "tests"); err == nil {
+		tests := make([]testing.InternalTest, 0)
+		internalTests := (*[]testing.InternalTest)(tPointer)
+		for _, test := range *internalTests {
+			td := &testDescriptor{
+				runner: r,
+				test:   test,
+				ran:    0,
+				failed: false,
+			}
+			tests = append(tests, testing.InternalTest{
+				Name: test.Name,
+				F:    td.run,
+			})
+		}
+		// Replace internal tests
+		*internalTests = tests
+	}
+}
+
+// Runs the test suite
+func (r *testRunner) Run() int {
+	return r.m.Run()
+}
+
+// Internal test runner, each test calls this method in order to handle retries and process exiting
+func (td *testDescriptor) run(t *testing.T) {
+	run := 1
+	maxRetries := td.runner.failRetriesCount
+	exitOnError := td.runner.exitOnError
+	var rc interface{}
+
+	for {
+		var innerTest *testing.T
+		title := "Run"
+		if run > 1 {
+			title = "Retry:" + strconv.Itoa(run-1)
+		}
+		title = "[runner." + title + "]"
+		t.Run(title, func(it *testing.T) {
+			defer func() {
+				rc = recover()
+				if rc != nil {
+					it.FailNow()
+				}
+			}()
+			innerTest = it
+			td.test.F(it)
+		})
+		innerTestInfo := getTestResultsInfo(innerTest)
+
+		if rc != nil {
+			if exitOnError {
+				panic(rc)
+			}
+			fmt.Println("PANIC RECOVER:", rc)
+			td.error = true
+		}
+		td.skipped = innerTestInfo.skipped
+		td.ran++
+
+		// Current run failure
+		if innerTestInfo.failed {
+			td.failed = true
+		}
+		// Current run ok but previous with fail -> Flaky
+		if !innerTestInfo.failed && td.failed {
+			td.failed = false
+			td.flaky = true
+			break
+		}
+
+		if td.skipped || !td.failed || run > maxRetries {
+			break
+		}
+		run++
+	}
+	if td.flaky {
+		fmt.Println("*** FLAKY", t.Name())
+	}
+	if td.error && exitOnError {
+		panic(rc)
+	}
+	if !td.error && !td.failed {
+		removeTestFailureFlag(t)
+	}
+}
+
+// Gets the test name
+func GetOriginalTestName(name string) string {
+	if runnerRegexName == nil {
+		runnerRegexName = regexp.MustCompile(`(?m)([\w -:_]*)\/\[runner.[\w:]*](\/[\w -:_]*)?`)
+	}
+	match := runnerRegexName.FindStringSubmatch(name)
+	if match == nil || len(match) == 0 {
+		return name
+	}
+	return match[1] + match[2]
+}
+
+func removeTestFailureFlag(t *testing.T) {
+	if t == nil {
+		return
+	}
+	if ptr, err := getFieldPointerOfT(t, "failed"); err == nil {
+		if *(*bool)(ptr) == true {
+			*(*bool)(ptr) = false
+			if parentPtr, err := getFieldPointerOfT(t, "parent"); err == nil {
+				parentTPointer := (**testing.T)(parentPtr)
+				if parentTPointer != nil && *parentTPointer != nil {
+					removeTestFailureFlag(*parentTPointer)
+				}
+			}
+		}
+	}
+}
```

Comment:
As discussed offline, this will mark the whole suite as passing whenever there is a flaky test, even if there were tests that previously failed

Reply:
right, I'm splitting this method in two, `setTestFailureFlag(t *testing.T, value bool)` and `getTestParent(t *testing.T) *testing.T`